### PR TITLE
Unittests

### PR DESCRIPTION
**Pull request by huangq1234553**
Created pull request to show code differences between implemented unittest framework and code in master branch. 

The test code has no dependencies on OpenFOAM or preCICE libraries (both are mocked via gmock / gtest).

To run the code:
- Build gmock and gtest (https://github.com/google/googletest/tree/master/googletest)
- Ensure the includes and built shared libraries are in CPLUS_INCLUDE_PATH and LD_LIBRARY_PATH respectively.
- Run the following commands:
```
cmake .
make -j <# procs>
./TestFolder/Test1
```
Probably a nice small exercise would be to write more tests for the configFileRead() method. You could make more .yml files that would test the different cases that could happen (i.e. tests to make sure that the wrong combination of inputs (i.e. FSI with faceNodes will return false) .

### CODE REVIEW

**Comment by MakisH** [comment]
With CMake 3.13.4 I get:

```
$ cmake .
CMake Warning at CMakeLists.txt:18 (project):
  VERSION keyword not followed by a value or was followed by a value that
  expanded to nothing.


-- The CXX compiler identification is GNU 5.4.0
-- The C compiler identification is GNU 5.4.0
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
CMake Error at /home/makish/inst/share/cmake-3.13/Modules/FindPackageHandleStandardArgs.cmake:137 (message):
  Could NOT find GTest (missing: GTEST_LIBRARY GTEST_INCLUDE_DIR
  GTEST_MAIN_LIBRARY)
Call Stack (most recent call first):
  /home/makish/inst/share/cmake-3.13/Modules/FindPackageHandleStandardArgs.cmake:378 (_FPHSA_FAILURE_MESSAGE)
  /home/makish/inst/share/cmake-3.13/Modules/FindGTest.cmake:196 (FIND_PACKAGE_HANDLE_STANDARD_ARGS)
  CMakeLists.txt:27 (find_package)


-- Configuring incomplete, errors occurred!
See also "/home/makish/adapter/Tests/UnitTests/CMakeFiles/CMakeOutput.log".
```

I have the following set up:
```bash
export CPLUS_INCLUDE_PATH="/home/makish/inst/googletest/googletest/include:/home/makish/inst/googletest/googlemock/include:$CPLUS_INCLUDE_PATH"
export LD_LIBRARY_PATH="/home/makish/inst/googletest/build/lib:$LD_LIBRARY_PATH"
```

I also tried with Google Test from the Ubuntu 16.04 repositories.

**Comment by huangq1234553** [comment]
Updated installation instructions:
To run the code:
Requires yaml-cpp-dev
'''
sudo apt install yaml-cpp-dev
'''

Download latest release:
https://github.com/google/googletest/releases
'''
tar xvf release-1.10.0.tar.gz && cd release-1.10.0
mkdir build && cd build
cmake -DBUILD_SHARED_LIBS=ON -DCMAKE_INSTALL_PREFIX=/path/to/gtest ..
make -j <# procs>
make install
'''

Run the following commands:
'''
export GTEST_ROOT=/path/to/gtest
cmake .
make -j <# procs>
./TestFolder/Test1
'''

**Comment by MakisH** [comment]
Now it configures and compiles, but it does not link:
```
[ 12%] Linking CXX executable TestFolder/Test1
/usr/bin/ld: cannot find -lyaml-cpp
/usr/bin/ld: cannot find -lgtest
/usr/bin/ld: cannot find -lgmock
collect2: error: ld returned 1 exit status
```

I also get the following warning:
```
-- Found GTest: /home/makish/inst/lib/libgtest.so  
CMake Warning (dev) at CMakeLists.txt:38 (add_executable):
  Policy CMP0037 is not set: Target names should not be reserved and should
  match a validity pattern.  Run "cmake --help-policy CMP0037" for policy
  details.  Use the cmake_policy command to set the policy and suppress this
  warning.

  The target name "TestFolder/Test1" is reserved or not valid for certain
  CMake features, such as generator expressions, and may result in undefined
  behavior.
This warning is for project developers.  Use -Wno-dev to suppress it.
```

I have set the following:
```
# YAML-CPP
export CPLUS_INCLUDE_PATH="/home/makish/inst/yaml-cpp/include" #:${CPLUS_INCLUDE_PATH}"
export LD_LIBRARY_PATH="/home/makish/inst/yaml-cpp/build:${LD_LIBRARY_PATH}"

# Google test
export GTEST_ROOT=/home/makish/inst
export GTEST_LIBRARY="$GTEST_ROOT/lib"
export GTEST_INCLUDE_DIR="$GTEST_ROOT/include"
export LD_LIBRARY_PATH="$GTEST_ROOT/lib:$LD_LIBRARY_PATH"
export CPLUS_INCLUDE_PATH="$GTEST_ROOT/include:$CPLUS_INCLUDE_PATH"


```

File: Tests/UnitTests/AdapterTest/AdapterTest.cc [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+#define private public
+
+#include "../../../Adapter.H"
+
+using namespace testing;
+
+const std::string INFO_LABEL = "info";
+const std::string ERROR_LABEL = "error-deferred";
+
+const std::string STANDARD_YAML_MESSAGE = "Reading the adapter's YAML configuration file "
```
Comment (MakisH) [inline Tests/UnitTests/AdapterTest/AdapterTest.cc]: We need to update this to the new preciceDict format.

**Comment by huangq1234553** [comment]
Using g++ 5.4.0. Had linking issues when yaml-cpp, gtest and gmock library (.so) files where in LD_LIBRARY_PATH. Added these paths to LIBRARY_PATH and compilation succeeded.

Need to investigate why this behavior is occuring?

**Comment by MakisH** [comment]
> Need to investigate why this behavior is occuring?

For now it's ok, let's just document this and focus on the actual tests.

**Comment by huangq1234553** [comment]
Created new branch for precice 2.0 api as it is easier to modify the updated openfoam adapter instead of merging.

See : https://github.com/precice/openfoam-adapter/pull/122

**Comment by MakisH** [comment]
@huangq1234553 is the branch [unittests](https://github.com/precice/openfoam-adapter/tree/unittests) still useful or can I delete it?

**Pull request closed without merging**